### DISCUSSION TITLE
Pylance with a multi-root workspace

### DISCUSSION THREAD

**Discussion by splint33** [microsoft/pylance-release#6303] [Q&A]
Hi everyone, I'm kinda new to Pylance, sorry if my question is dumb.
I have a VSCode workspace, containing multiple folders. Each folder is a GIT repository.
My code-workspace is looking like that at the moment:

```
{
    "folders": [
        {
            "name": "root",
            "path": "."
        },
        {
            "name": "lib_1",
            "path": "C:\\AutomatedTests\\lib_1"
        },
        {
            "name": "lib_2",
            "path": "C:\\AutomatedTests\\lib_2"
        },
        {
            "name": "lib_3",
            "path": "C:\\AutomatedTests\\lib_3"
        }
    ],
    "settings": {
        "python.analysis.extraPaths": [
            "${workspaceFolder:root}/scripts",
            "${workspaceFolder:lib_1}",
            "${workspaceFolder:lib_2}",
            "${workspaceFolder:lib_3}"
        ]
}
```

When I edit lib1, I must restart the language server in order to be able to see changes in the root folder.

I understand that Pylance is indexing each folder individually at the start of the language server, and seems to use a cached version of the index.

Changes made into lib1 are reflected into lib1 immediately, but not on other workspace folders.
Is there a way to keep this structure and make it work with pylance ?
Thanks

**Comment by heejaechang** [comment]
can you tell me what you are looking for with this?

```
"python.analysis.extraPaths": [
            "${workspaceFolder:root}/scripts",
            "${workspaceFolder:lib_1}",
            "${workspaceFolder:lib_2}",
            "${workspaceFolder:lib_3}"
        ]
```

in `multiroot workspace`, `${workspaceFolder:xxxx}` in `extraPaths` is used to indicate which workspace the option is applied to. basically, `root` will only see `/scripts` and `lib_1` will only see `<empty>` as extra path.

if your intension is making `lib_x` and `scripts` to be accessible for all workspaces, you should write like this

```
"python.analysis.extraPaths": [
            "./scripts",
            "C:\\AutomatedTests\\lib_1",
           ...
        ]
```

for this, `When I edit lib1, I must restart the language server in order to be able to see changes in the root folder.`, what you mean by `see changes`?

**Comment by splint33** [comment]
Without the extraPath, Pylance will not resolve the imports on my project.
I have tried to replace the extraPath by absolute path, without using ${workspaceFolder:xxx} format, but I face the same problem.

By see changes, I mean:
I am working on the "root" project, importing and using lib_1.
I am facing a case when lib_1 is missing a method on a class, so I implement a new method on the class.
When I try to use the newly created class method on "root", Pylance is not seeing it, as if I was calling an non existing method.
I must restart the language server, or restart VSCode entierly, in order to have Pylance finding the newly created class method on the root folder.

Thanks for your time

**Reply by heejaechang** [reply]
did you change the file in `lib_1` in `vscode`? or outside of `vscode`?

**Reply by heejaechang** [reply]
I can repro it with file being changed in vscode. let me take a look.

**Reply by splint33** [reply]
I am changing the file inside vscode.
Thanks for taking a look !

**Reply by heejaechang** [reply]
issue tracking - https://github.com/microsoft/pylance-release/issues/6318
